services/statestorage/adapter/memory_persistence.go

```diff
@@ -4,41 +4,79 @@ import (
 	"github.com/orbs-network/orbs-spec/types/go/primitives"
 	"github.com/orbs-network/orbs-spec/types/go/protocol"
 	"github.com/pkg/errors"
-)
-
-type Config interface {
-}
+	)
 
 type InMemoryStatePersistence struct {
-	stateWritten chan bool
-	stateDiffs   map[primitives.ContractName]map[string]*protocol.StateRecord
-	config       Config
+	stateWritten 	chan bool
+	stateDiffs   	map[primitives.BlockHeight]map[primitives.ContractName]map[string]*protocol.StateRecord
 }
 
-func NewInMemoryStatePersistence(config Config) StatePersistence {
+func NewInMemoryStatePersistence() StatePersistence {
+	stateDiffsContract :=  map[primitives.ContractName]map[string]*protocol.StateRecord{primitives.ContractName("BenchmarkToken"): {}}
+
 	return &InMemoryStatePersistence{
-		config: config,
 		// TODO remove init with a hard coded contract once deploy/provisioning of contracts exists
-		stateDiffs:   map[primitives.ContractName]map[string]*protocol.StateRecord{primitives.ContractName("BenchmarkToken"): {}},
+		stateDiffs:  map[primitives.BlockHeight]map[primitives.ContractName]map[string]*protocol.StateRecord{primitives.BlockHeight(0): stateDiffsContract},
 		stateWritten: make(chan bool, 10),
 	}
 }
 
-func (sp *InMemoryStatePersistence) WriteState(contract primitives.ContractName, stateDiff *protocol.StateRecord) error {
-	if _, ok := sp.stateDiffs[contract]; !ok {
-		sp.stateDiffs[contract] = map[string]*protocol.StateRecord{}
+func (sp *InMemoryStatePersistence) WriteState(height primitives.BlockHeight, contract primitives.ContractName, stateDiff *protocol.StateRecord) error {
+	if _, ok := sp.stateDiffs[height]; !ok {
+		sp.stateDiffs[height] = sp.cloneCurrentStateDiff(height)
 	}
 
-	sp.stateDiffs[contract][stateDiff.Key().KeyForMap()] = stateDiff
+	if _, ok := sp.stateDiffs[height][contract]; !ok {
+		sp.stateDiffs[height][contract] = map[string]*protocol.StateRecord{}
+	}
+
+	sp.stateDiffs[height][contract][stateDiff.Key().KeyForMap()] = stateDiff
+
 	sp.stateWritten <- true
 
 	return nil
 }
 
-func (sp *InMemoryStatePersistence) ReadState(contract primitives.ContractName) (map[string]*protocol.StateRecord, error){
-	if contractStateDiff, ok := sp.stateDiffs[contract]; ok {
-		return contractStateDiff, nil
+func (sp *InMemoryStatePersistence) cloneCurrentStateDiff(height primitives.BlockHeight) map[primitives.ContractName]map[string]*protocol.StateRecord {
+	prevHeight := height - primitives.BlockHeight(1)
+	if _, ok := sp.stateDiffs[prevHeight]; !ok {
+		panic ("trying to commit blocks not in order")
```

Comment:
I think we decided to avoid all `panics` in production code because we can't really handle the node crashing. A better alternative would be to log an error to reporting, and recover gracefully